### PR TITLE
Pull request for #205

### PR DESCRIPTION
**Pull request by matevzmihalic**
I added custom authorization, that hides non published posts from other users.

### CODE REVIEW

**Comment by mitar** [comment]
Without tests this is useless. First make tests which shows scenarios we want to allow and scenarios we do not want. So we want users to be able to post new posts, only authors should be able to add attachments, only authors should be able to modify anything before publishing. Everybody should be able to comment. Only authors of comments and post owner should be able to delete the comment. Nobody should be allowed to change their comments. Before publishing, author can change the post. After publishing nobody can change the post anymore.

I think we want to define also our own authentication class, which allows everything, but for `get_identifier` returns username.

**Comment by mitar** [comment]
Ah, sorry. I was thinking that you are doing the authorization for what we discussed at the class. Of all necessary checks we still have to do.

OK, if this is just for #205, then it looks good enough. But tests are still missing.

**Comment by mitar** [comment]
I will open another ticket for the rest.

File: piplmesh/api/resources.py [hunk 1]
```diff
@@ -5,6 +5,13 @@
 from piplmesh.account import models as account_models
 from piplmesh.api import models as api_models
 
+class PiplMeshAuthorization(tastypie_authorization.Authorization):
```
Comment (mitar) [inline piplmesh/api/resources.py]: Please move this to another Python module.

File: piplmesh/api/resources.py [hunk 2]
```diff
@@ -30,7 +37,7 @@ class Meta:
         object_class = api_models.Comment
         allowed_methods = ('get', 'post', 'put', 'patch', 'delete')
         # TODO: Make proper authorization, current implementation is for development use only
-        authorization = tastypie_authorization.Authorization()
+        authorization = PiplMeshAuthorization()
```
Comment (mitar) [inline piplmesh/api/resources.py]: This is not good. User does not have `is_published`.

File: piplmesh/api/resources.py [hunk 3]
```diff
@@ -51,7 +58,7 @@ class Meta:
         object_class = api_models.Attachment
         allowed_methods = ('get', 'post', 'put', 'patch', 'delete')
         # TODO: Make proper authorization, current implementation is for development use only
-        authorization = tastypie_authorization.Authorization()
+        authorization = PiplMeshAuthorization()
```
Comment (mitar) [inline piplmesh/api/resources.py]: This is not good. Attachment does not have `is_published` flag.

File: piplmesh/api/resources.py [hunk 4]
```diff
@@ -1,9 +1,9 @@
-from tastypie import authorization as tastypie_authorization, fields as tastypie_fields
+from tastypie import fields as tastypie_fields
 
 from tastypie_mongoengine import fields, resources
 
 from piplmesh.account import models as account_models
-from piplmesh.api import models as api_models
+from piplmesh.api import authorization, models as api_models
```
Comment (mitar) [inline piplmesh/api/resources.py:7]: Please split this into multiple lines. You already have if-block. So just continue using it.

Comment (matevzmihalic) [inline piplmesh/api/resources.py:7]: You mean like this:
`from piplmesh.api import authorization`
`from piplmesh.api import models as api_models`

**Comment by mitar** [comment]
I moved above comment to #231.

File: piplmesh/api/resources.py (hunk 4 above)
Comment (mitar) [inline piplmesh/api/resources.py:7]: UPS. This is completely wrong space for this comment. I wonder why is here??? You made a change while I was commenting. ;-)

This is for the if-block inside the authorization class. Where you have a inline if-then-else block.

Comment (mitar) [inline piplmesh/api/resources.py:7]: (The import here is of course correct. Don't change anything.)

File: piplmesh/api/authorization.py [hunk 5]
```diff
@@ -0,0 +1,10 @@
+from tastypie import authorization as tastypie_authorization
+
+class PiplMeshAuthorization(tastypie_authorization.Authorization):
+    def is_authorized(self, request, object=None):
+        try:
+            obj = self.resource_meta.queryset[0]
+        except (IndexError, TypeError):
+            return True
+        
+        return True if obj.is_published else obj.author == request.user
```
Comment (mitar) [inline piplmesh/api/authorization.py]: This. This is unnecessary to do inline.

File: piplmesh/api/authorization.py [hunk 6]
```diff
@@ -0,0 +1,10 @@
+from tastypie import authorization as tastypie_authorization
+
+class PiplMeshAuthorization(tastypie_authorization.Authorization):
+    def is_authorized(self, request, object=None):
+        try:
+            obj = self.resource_meta.queryset[0]
```
Comment (mitar) [inline piplmesh/api/authorization.py]: Why is this needed and `object` is not good enough?

File: piplmesh/api/resources.py [hunk 7]
```diff
@@ -45,13 +46,14 @@ class LinkAttachmentResource(AuthoredResource):
...
         object_class = api_models.LinkAttachment
+        authorization = authorization.PiplMeshAuthorization()
 
 class AttachmentResource(AuthoredResource):
     class Meta:
         object_class = api_models.Attachment
         allowed_methods = ('get', 'post', 'put', 'patch', 'delete')
         # TODO: Make proper authorization, current implementation is for development use only
-        authorization = tastypie_authorization.Authorization()
+        authorization = authorization.PiplMeshAuthorization()
```
Comment (mitar) [inline piplmesh/api/resources.py]: Please see my comments from before. Attachments do not have is published flag?

File: piplmesh/api/authorization.py (hunk 6 above)
Comment (matevzmihalic) [inline piplmesh/api/authorization.py]: I don't know what's up with object, but it's always None ;)

File: piplmesh/api/resources.py (hunk 7 above)
Comment (matevzmihalic) [inline piplmesh/api/resources.py]: Just commited this.

File: piplmesh/api/authorization.py (hunk 6 above)
Comment (mitar) [inline piplmesh/api/authorization.py]: A bug in our Tastypie implementation? Don't workaround. Fix (or at least make a test there which fails and I will fix it).

File: piplmesh/api/resources.py [hunk 8]
```diff
@@ -45,6 +46,7 @@ class LinkAttachmentResource(AuthoredResource):
 
     class Meta:
         object_class = api_models.LinkAttachment
+        authorization = authorization.PiplMeshAuthorization()
```
Comment (mitar) [inline piplmesh/api/resources.py]: Ehm. This does not exist anymore. And this line you can remove, it was not here before.

File: piplmesh/api/resources.py [hunk 9]
```diff
@@ -38,6 +38,7 @@ class ImageAttachmentResource(AuthoredResource):
 
     class Meta:
         object_class = api_models.ImageAttachment
+        authorization = tastypie_authorization.Authorization()
```
Comment (mitar) [inline piplmesh/api/resources.py]: This line you can remove, it was not here before.

File: piplmesh/api/authorization.py (hunk 6 above)
Comment (matevzmihalic) [inline piplmesh/api/authorization.py]: Not our bug, that is just how tastypie works (currently object is always none; nowhere in tastypies code is is_authorized called with object argument).
I replaced is_authorized with apply_limits which is more suited for this kind of work.

File: piplmesh/api/authorization.py [hunk 10]
```diff
@@ -0,0 +1,11 @@
+from mongoengine import queryset
+from tastypie import authorization as tastypie_authorization
+
+class PostAuthorization(tastypie_authorization.Authorization):
+    def apply_limits(self, request, object_list):
+        if request and hasattr(request, 'user'):
```
Comment (mitar) [inline piplmesh/api/authorization.py:6]: Why would `request` and request user not be available? When you user OAuth? But we will have to change this limits anyway then.

File: piplmesh/api/authorization.py [hunk 11]
```diff
@@ -0,0 +1,11 @@
...
+            if request.user.is_authenticated():
```
Comment (mitar) [inline piplmesh/api/authorization.py]: This check is not needed anymore. All our users are always real users, lazy or non-lazy. We are misusing `is_authenticated` now for telling between lazy and non-lazy users. So this is not needed here.

We should just apply filter so that only published and only those authored are visible. The query can be simplified accordingly. Think about it.

File: piplmesh/api/authorization.py (hunk 10 above)
Comment (matevzmihalic) [inline piplmesh/api/authorization.py:6]: Don't know, I just copied it from documentation. If it's in there it should be right ;-)

File: piplmesh/api/authorization.py (hunk 11 above)
Comment (matevzmihalic) [inline piplmesh/api/authorization.py]: Yes, this is not needed, I will remove it.

File: piplmesh/api/authorization.py (hunk 10 above)
Comment (mitar) [inline piplmesh/api/authorization.py:6]: See in the code when request is not passed in. Because sometimes we could maybe rather fail then return unrestricted access.

(Or return simply empty list if request is not available.)

(So the difference here is between deny by default or allow by default.)

Comment (matevzmihalic) [inline piplmesh/api/authorization.py:6]: I decided to return only published posts when request or request.user aren't available.

Comment (mitar) [inline piplmesh/api/authorization.py:6]: Good decision!

File: piplmesh/api/authorization.py [hunk 12]
```diff
@@ -0,0 +1,11 @@
+from mongoengine import queryset
+from tastypie import authorization as tastypie_authorization
+
+class PostAuthorization(tastypie_authorization.Authorization):
+    def apply_limits(self, request, object_list):
+        if request and hasattr(request, 'user'):
+            object_list = object_list.filter((queryset.Q(author=request.user) & queryset.Q(is_published=False)) | queryset.Q(is_published=True))
```
Comment (mitar) [inline piplmesh/api/authorization.py]: This can be simplified! ;-)

Comment (matevzmihalic) [inline piplmesh/api/authorization.py]: Done!

**Comment by mitar** [comment]
Good!
